pymatgen/io/core.py

```diff
@@ -0,0 +1,240 @@
+            make_dir: Whether to create the directory if it does not already exist.
+            overwrite: Whether to overwrite an input file if it already exists.
+            Additional kwargs are passed to generate_inputs
+            zip_inputs: If True, inputs will be zipped into a file with the
+                same name as the InputSet (e.g., InputSet.zip)
+        """
+        path = directory if isinstance(directory, Path) else Path(directory)
+
+        for fname, contents in self._inputs.items():
+            file = path / fname
+
+            if not path.exists():
+                if make_dir:
+                    path.mkdir(parents=True, exist_ok=True)
+
+            if file.exists() and not overwrite:
+                raise FileExistsError(f"File {str(fname)} already exists!")
+            file.touch()
+
+            # write the file
+            if isinstance(contents, str):
+                with zopen(file, "wt") as f:
+                    f.write(contents)
+            else:
+                contents.write_file(file)
+
+        if zip_inputs:
+            zipfilename = path / f"{self.__class__.__name__}.zip"
+            with ZipFile(zipfilename, "w") as zip:
+                for fname, contents in self._inputs.items():
+                    file = path / fname
+                    try:
+                        zip.write(file)
+                        os.remove(file)
+                    except FileNotFoundError:
+                        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def from_directory(cls, directory: Union[str, Path]):
+        """
+        Construct an InputSet from a directory of one or more files.
+
+        Args:
+            directory: Directory to read input files from
+        """
+        raise NotImplementedError(f"from_directory has not been implemented in {cls}")
+
+    def validate(self) -> bool:
+        """
+        A place to implement basic checks to verify the validity of an
+        input set. Can be as simple or as complex as desired.
+
+        Will raise a NotImplementedError unless overloaded by the inheriting class.
+        """
+        raise NotImplementedError(f".validate() has not been implemented in {self.__class__}")
+
+    def __len__(self):
+        return len(self._inputs.keys())
+
+    def __iter__(self):
+        return iter(self._inputs.items())
+
+    def __getitem__(self, key):
+        return self._inputs[key]
+
+
+class InputSetGenerator(MSONable):
+    """
+    InputSetGenerator classes serve as generators for Input objects. They contain
+    settings or sets of instructions for how to create Input from a set of
+    coordinates or a previous calculation directory.
+    """
+
+    @abc.abstractmethod
+    def get_input_set(self) -> InputSet:
+        """
+        Generate an InputSet object. Typically the first argument to this method
+        will be a Structure or other form of atomic coordinates.
+        """
+        pass
+
+
+class TemplateInputSet(InputSet):
```

Comment:
This definitely shouldn't be in core. If it is for testing purposes, put it in the test itself for now. When there is an actual application, you can then move it.

Reply:
OK. I do intend this class to be useful so I've moved it to a new file called `template.py`. This is the minimal / lowest-barrier way to transition from manual preparation of input files to using pymatgen to manage calc inputs.